src/bootstrap/native.rs

```diff
@@ -389,12 +393,82 @@ impl Step for Llvm {
 
         cfg.build();
 
+        if builder.config.bolt_profile_generate {
+            instrument_llvm_with_bolt(builder);
+        }
+        if let Some(path) = &builder.config.bolt_profile_use {
+            apply_bolt_profile_to_llvm(path);
+        }
+
         t!(stamp.write());
 
         build_llvm_config
     }
 }
 
+fn apply_bolt_profile_to_llvm(profile_path: &str) {
+    let lib_dir = PathBuf::from("/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/lib");
+    let llvm_lib_path = lib_dir.join("libLLVM-14-rust-1.61.0-nightly.so");
+    // The library cannot be instrumented in-place
+    let llvm_optimized_lib_path = PathBuf::from("/tmp/libLLVM-optimized.so");
+
+    let status = Command::new("llvm-bolt")
+        .arg(&llvm_lib_path)
+        .arg("-data")
+        .arg(&profile_path)
+        .arg("-o")
+        .arg(&llvm_optimized_lib_path)
+        .arg("-reorder-blocks")
+        .arg("cache+")
+        .arg("-reorder-functions")
+        .arg("hfsort")
+        .arg("-split-functions")
+        .arg("2")
+        .arg("-split-all-cold")
+        .arg("-split-eh")
+        .arg("-dyno-stats")
```

Comment:
nit: it looks like currently this isn't updating debuginfo mappings, which we'll probably want to do. The bolt readme currently discusses being incompatible with dwarf v5; not sure how that plays into things (https://github.com/llvm/llvm-project/blob/main/bolt/README.md#input-binary-requirements). Maybe we should pass the flag they suggest while building LLVM.

(Alternatively, it might be feasible to delete debuginfo from LLVM specifically via strip or something...)

Locally I'm seeing:

````
### unhandled dwarf2 abbrev form code 0x25
### unhandled dwarf2 abbrev form code 0x25
### unhandled dwarf2 abbrev form code 0x25
### unhandled dwarf2 abbrev form code 0x23
### unhandled dwarf2 abbrev form code 0x25
### unhandled dwarf2 abbrev form code 0x25
### unhandled dwarf2 abbrev form code 0x25
### unhandled dwarf2 abbrev form code 0x23
==1906996== Valgrind: debuginfo reader: ensure_valid failed:
==1906996== Valgrind:   during call to ML_(img_get)
==1906996== Valgrind:   request for range [1903637143, +4) exceeds
==1906996== Valgrind:   valid image size of 1937824 for image:
==1906996== Valgrind:   "/home/mark/.rustup/toolchains/e4a59c8375b06f49413c0664fc3c62a2084ffd18/bin/rustc"
==1906996==
==1906996== Valgrind: debuginfo reader: Possibly corrupted debuginfo file.
==1906996== Valgrind: I can't recover.  Giving up.  Sorry.
==1906996==
```

when trying to pass the built binary to cachegrind.

Reply:
I used `readelf --debug-dump=info rustc | grep -A 2 'Compilation Unit @'` on bootstrap `rustc` and it seems to me that we already use DWARF version 4. I will add the `-update-debug-sections` flag to BOLT to keep debug information.

Reply:
Clang switched to Dwarf5 : https://reviews.llvm.org/rGd3b26dea1610

Reply:
Just a reminder: bolt now supports DWARF 5: https://github.com/llvm/llvm-project/commit/014cd37f5141af24c3082c70a251f84f48992022 https://github.com/llvm/llvm-project/commit/4a58eb9e4e5282df95deff243f4141b35909710c